Add each fronius sensor as own template

nielstron opened the pull request:
## Breaking Change:
## Description:
This update adds the often used templates to the state engine per default and hides the overall components. The reason is that there is almost no use case without the templates and they are tricky yet repetitive to define.

**Related issue (if applicable):** as an example [issue 11 on the libraries page](https://github.com/nielstron/pyfronius/issues/11)

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io):**  home-assistant/home-assistant.io#10028

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

nielstron committed:
Add each sensor as own template

nielstron committed:
Fix formatting in history test

MartinHjelmare commented on homeassistant/components/fronius/sensor.py:
```diff
@@ -94,22 +95,31 @@ def _device_id_validator(config):
     """The Fronius sensor implementation."""
 
-    def __init__(self, data, name, device):
+    def __init__(self, data, name, device, add_entities):
```
The entity shouldn't be responsible for adding other entities.

MartinHjelmare commented on homeassistant/components/fronius/sensor.py:
```diff
@@ -137,13 +147,27 @@ def device_state_attributes(self):
             for key in values:
                 if 'value' in values[key]:
-                    attributes[key] = values[key].get('value', 0)
+                    attributes[key] = values[key]
```
If we add sensors for the keys we shouldn't add them as state attributes on this entity.

MartinHjelmare commented on homeassistant/components/fronius/sensor.py:
```diff
@@ -94,22 +95,31 @@ def _device_id_validator(config):
+        self._add_entities = add_entities
+
+    @property
+    def hidden(self):
```
Please remove this.

MartinHjelmare reviewed (changes requested):
If we need a manager of entities that manager shouldn't be an entity itself.

nielstron committed:
Make adapters non-entities

nielstron committed:
Externalize and organize data fetching, improve system view

MartinHjelmare commented on homeassistant/components/fronius/sensor.py:
```diff
@@ -68,7 +75,7 @@ def _device_id_validator(config):
     fronius = Fronius(session, config[CONF_RESOURCE])
 
-    sensors = []
+    scan_interval = config[CONF_SCAN_INTERVAL]
```
We can use dict.get here to set a default interval.

nielstron replied:
Okay. It just confused me that `get` would return None if no scan interval is given. I somehow tried to get a system wide set default scan interval at first but that doesn't seem to exist.
Thanks!

MartinHjelmare commented on homeassistant/components/fronius/sensor.py:
```diff
@@ -136,62 +150,139 @@ def device_state_attributes(self):
+
 
-class FroniusInverterSystem(FroniusSensor):
+class FroniusInverterSystem(FroniusAdapter):
```
Please don't have the entities inherit the adapter. Please separate the entity from the instance that handles the device data and updates.

MartinHjelmare replied:
I see these are not entities anymore. But it's confusing since we call them sensor classes and keep some entity specific properties like `device_state_attributes`. Maybe it's enough to rename some things about these to make it clear that they are not entities?

nielstron replied:
I consistently renamed them adapters now, that should make things clearer.

MartinHjelmare replied:
Great! Thanks!

nielstron committed:
Small fixes

Rename fetching adapters to adapters
throw away non-working system overviews
slightly change naming
remove scan_interval from schema
formatting

MartinHjelmare commented on homeassistant/components/fronius/sensor.py:
```diff
@@ -68,62 +73,69 @@ def _device_id_validator(config):
 
-    async_add_entities(sensors, True)
+        async def fetch_data(*_):
+            return await adapter.async_update()
```
Nothing is checking this return value.

MartinHjelmare reviewed (approved):
Looks good! Can be merged when last comment is addressed and build passes.

nielstron committed:
Scan interval is already timedelta and unnecessary return

MartinHjelmare commented:
Please rebase to solve the merge conflict.

nielstron committed:
Merge remote-tracking branch 'ha/dev' into pyfronius

nielstron committed:
Formatting

nielstron committed:
Ensure better codestyle by storing cell variables explicitely in different places
